BGP CP: Replaces LocalNodeStore with Local CiliumNode

danehans opened the pull request:
- Replaces the LocalNodeStore with the existing local CiliumNode throughout the BGP CP.
- Updates unit and integration tests accordingly.
- Adds the GetIP() method for returning the CiliumNode IP.

Fixes: #28237

harsimran-pabla commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -465,3 +466,37 @@ func (n *CiliumNode) InstanceID() (instanceID string) {
+// - NodeInternalIP
+// - NodeExternalIP
+// - other IP address type
+func (n *CiliumNode) GetIP(ipv6 bool) net.IP {
```
Consider changing net.IP to netip.Addr.

danehans replied:
Commit bb0f39b resolves this comment.

harsimran-pabla commented on pkg/bgpv1/test/adverts_test.go:
```diff
@@ -155,17 +150,20 @@ func Test_PodCIDRAdvert(t *testing.T) {
+	obj, err := tracker.Get(v2.SchemeGroupVersion.WithResource("ciliumnodes"), "", baseNode.name)
+	require.NoError(t, err)
+	node, ok := obj.(*v2.CiliumNode)
+	require.Equal(t, ok, true)
```
nit: could be require.True.

danehans replied:
Commit bb0f39b resolves this comment.

rastislavs reviewed (approved):
LGTM

gandro commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -465,3 +467,43 @@ func (n *CiliumNode) InstanceID() (instanceID string) {
 	return
 }
+
+// GetIP returns one of the CiliumNode's IP addresses available with the
```
This duplicates a lot of existing logic and I'm a bit worried that the two implementations will eventually diverge if future contributions touch up one, but not the other. We should at least document that this mirrors the logic `GetNodeIP` of `node/types.Node`, or maybe even better try to have both implementations use the same "ip selection logic".


We could probably do that by defining an interface as follow that is implemented for both [v2.NodeAddress](https://github.com/cilium/cilium/blob/v1.14.2/pkg/k8s/apis/cilium.io/v2/types.go#L334) and [types.Address](https://github.com/cilium/cilium/blob/762d8c9ed13d4d2bced71d4b04c1a30de149bf93/pkg/node/types/node.go#L285), e.g.

```go
type NodeAddress interface {
   Type() addressing.AddressType
   Addr() netip.Addr
}

func ExtractNodeIP[T NodeAddress](addrs []T, ipv6 bool) (*netip.Addr, error) {
   // ...
}
```

And then have both `CiliumNode.GetIP` and `Node.GetNodeIP` call into `ExtractNodeIP` to use the same logic.

danehans replied:
@gandro thanks for the review and guidance. Commit [fbeff7b](https://github.com/cilium/cilium/pull/28238/commits/e6c350990432203bbd2c72b0032a5e2f7fc0f37a) implements ^ comment.

gandro reviewed (commented):
This is a great simplification! Looks good to me in terms of logic, but I do have a concern about logic duplication

gandro commented on pkg/node/addressing/addresstype.go:
```diff
@@ -16,3 +21,47 @@ const (
+	ToString() string
+}
+
+// GetIP returns one of the provided IP addresses available with the following priority:
```
nit:

```suggestion
// ExtractNodeIP returns one of the provided IP addresses available with the following priority:
```

danehans replied:
Thanks for catching this mistake @gandro. Commit 4a735bb resolves your comment.

gandro reviewed (approved):
Thanks!

danehans committed:
pkg/node: Abstracts GetNodeIP() Method

Abstracts the GetNodeIP() method for use by other types.

Signed-off-by: Daneyon Hansen <daneyon.hansen@solo.io>

danehans committed:
BGP CP: Replaces LocalNodeStore with Local CiliumNode

- Replaces the LocalNodeStore with the existing local CiliumNode
  throughout the BGP CP.
- Updates unit and integration tests accordingly.
- Adds the GetIP() method for returning the CiliumNode IP.

Signed-off-by: Daneyon Hansen <daneyon.hansen@solo.io>
